inc/sitemaps/class-sitemaps.php

```diff
@@ -0,0 +1,429 @@
+
+		if ( $type === '1' ) {
+			$this->build_root_map();
+			return;
+		}
+
+		foreach ( $this->providers as $provider ) {
+			if ( ! $provider->handles_type( $type ) ) {
+				continue;
+			}
+
+			$links = $provider->get_sitemap_links( $type, $this->max_entries, $this->current_page );
+
+			if ( empty( $links ) ) {
+				$this->bad_sitemap = true;
+
+				return;
+			}
+
+			$this->sitemap = $this->renderer->get_sitemap( $links, $type, $this->current_page );
+
+			return;
+		}
+
+		if ( has_action( 'wpseo_do_sitemap_' . $type ) ) {
+			do_action( 'wpseo_do_sitemap_' . $type );
+			return;
+		}
+
+		$this->bad_sitemap = true;
+	}
+
+	/**
+	 * Build the root sitemap (example.com/sitemap_index.xml) which lists sub-sitemaps for other content types.
+	 */
+	public function build_root_map() {
+
+		$links = array();
+
+		foreach ( $this->providers as $provider ) {
+			$links = array_merge( $links, $provider->get_index_links( $this->max_entries ) );
+		}
+
+		$this->sitemap = $this->renderer->get_index( $links );
+	}
+
+	/**
+	 * Function to dynamically filter the change frequency.
+	 *
+	 * @param string $filter  Expands to wpseo_sitemap_$filter_change_freq, allowing for a change of the frequency for numerous specific URLs.
+	 * @param string $default The default value for the frequency.
+	 * @param string $url     The URL of the current entry.
+	 *
+	 * @return mixed|void
+	 */
+	static public function filter_frequency( $filter, $default, $url ) {
+		/**
+		 * Filter: 'wpseo_sitemap_' . $filter . '_change_freq' - Allow filtering of the specific change frequency
+		 *
+		 * @api string $default The default change frequency
+		 */
+		$change_freq = apply_filters( 'wpseo_sitemap_' . $filter . '_change_freq', $default, $url );
+
+		if ( ! in_array( $change_freq, array(
+			'always',
+			'hourly',
+			'daily',
+			'weekly',
+			'monthly',
+			'yearly',
+			'never',
+		) )
+		) {
+			$change_freq = $default;
+		}
+
+		return $change_freq;
+	}
+
+	/**
+	 * Spits out the XSL for the XML sitemap.
+	 *
+	 * @param string $type Type to output.
+	 *
+	 * @since 1.4.13
+	 */
+	public function xsl_output( $type ) {
+
+		if ( $type !== 'main' ) {
+
+			do_action( 'wpseo_xsl_' . $type ); // TODO document action. R.
+
+			return;
+		}
+
+		header( $this->http_protocol . ' 200 OK', true, 200 );
+		// Prevent the search engines from indexing the XML Sitemap.
+		header( 'X-Robots-Tag: noindex, follow', true );
+		header( 'Content-Type: text/xml' );
+
+		// Make the browser cache this file properly.
+		$expires = YEAR_IN_SECONDS;
+		header( 'Pragma: public' );
+		header( 'Cache-Control: maxage=' . $expires );
+		header( 'Expires: ' . gmdate( 'D, d M Y H:i:s', ( time() + $expires ) ) . ' GMT' );
+
+		require_once( WPSEO_PATH . 'css/xml-sitemap-xsl.php' );
+	}
+
+	/**
+	 * Spit out the generated sitemap and relevant headers and encoding information.
+	 */
+	public function output() {
+
+		if ( ! headers_sent() ) {
+			header( $this->http_protocol . ' 200 OK', true, 200 );
```

Comment:
I didn't feel like headers are part of rendering (producing XML string in other words), but haven't finalized my opinion on it.